### PR TITLE
fix(testing): reset collected statuses in Harness.evaluate_status

### PR DESCRIPTION
**Pull request by benhoyt**
Currently you can't really call `Harness.evaluate_status()` twice in one test because it keeps the previous collected statuses around. Reset them before evaluating.

I believe this addresses https://github.com/canonical/operator/issues/736#issuecomment-1762094066

### CODE REVIEW

**Review by tonyandrewmeyer — APPROVED** [review]

File: test/test_testing.py [hunk 1]
```diff
@@ -2828,20 +2828,28 @@ def __init__(self, framework: ops.Framework):
...
-        self.assertEqual(harness.model.app.status, ops.ActiveStatus())
-        self.assertEqual(harness.model.unit.status, ops.BlockedStatus('bar'))
+        self.assertEqual(harness.model.app.status, ops.BlockedStatus('blocked app'))
+        self.assertEqual(harness.model.unit.status, ops.BlockedStatus('blocked unit'))
+
+        harness.charm.app_status_to_add = ops.ActiveStatus('active app')
+        harness.charm.unit_status_to_add = ops.ActiveStatus('active unit')
+        harness.evaluate_status()
+        self.assertEqual(harness.model.app.status, ops.ActiveStatus('active app'))
+        self.assertEqual(harness.model.unit.status, ops.ActiveStatus('active unit'))
```
Comment (sed-i) [inline test/test_testing.py:2852]: So evaluate_status calculates, commit somewhere else, and clears the collection for the next add_status?
Then should probably document that cannot do "additive tests" with evaluate_status.

Not sure which implementation is less confusing though.

File: ops/testing.py [hunk 2]
```diff
@@ -1634,6 +1634,9 @@ def evaluate_status(self) -> None:
         Tests should normally call this and then assert that ``self.model.app.status``
         or ``self.model.unit.status`` is the value expected.
         """
+        if self._backend._is_leader:
+            self.charm.app._collected_statuses = []
```
Comment (sed-i) [inline ops/testing.py:1638]: The distinction for leader seems inconsistent with the `clear`:
If `evaluate_status` clears the stage for the next `add_status`, it's a harness-level function. Leadership status is on a different level.

File: test/test_testing.py (hunk 1 above)
Comment (benhoyt) [reply test/test_testing.py:2852]: Good call. I've updated the `evaluate_status` docstring to clarify that. I don't think "additive tests" is useful for this: before triggering the collect-status events, the statuses added should be reset.

File: ops/testing.py (hunk 2 above)
Comment (benhoyt) [reply ops/testing.py]: Yeah, good call. I've now cleared this independent of `is_leader`.

**Pull request merged**